### PR TITLE
Username and password support for HttpGitClient

### PR DESCRIPTION
**Pull request by mjmaenpaa**
Added tests for original pull request #404 by @jsbain and rebased jsbain's changes against master. Original pull request fixes issue #249. Additionally added support for only supplying username which allows to use github personal access tokens in url. This should fix issue #428.

I wrote few tests for this that make sure that username and password end up in HttpGitClient when using get_transport_and_path. Additionally tests make sure that init actually ends up calling HTTPPasswordMgrWithDefaultRealm.add_password.

### CODE REVIEW

**Comment by gadams999** [comment]
Any update on when these pulls will be accepted? I have a use case where username:password is needed (AWS Lambda function to get a ref and create tgz from CodeCommit). Using the GitHub version, seeing the same `httplib.InvalidURL: nonnumeric port` error.

Loving that dulwich is completely native as there are not git clients in Lambda.

**Comment by jelmer** [comment]
I was going to make changes to this pull request while merging it, but I haven't found time to do so. I'll comment here instead - if somebody can address these issues, that would be great.

File: .travis.yml [hunk 1]
```diff
@@ -5,9 +5,9 @@ env:
 matrix:
   include:
     - python: "2.7"
-      env: TEST_REQUIRE="gevent greenlet geventhttpclient fastimport"
+      env: TEST_REQUIRE="gevent greenlet geventhttpclient fastimport mock"
```
Comment (jelmer) [inline .travis.yml]: Please don't add extra dependencies. There should be no need for mocking here; you can just create a HttpGitClient and verify that the right attributes are set.

File: dulwich/client.py [hunk 2]
```diff
@@ -997,13 +997,22 @@ def default_urllib2_opener(config):
 
 class HttpGitClient(GitClient):
 
-    def __init__(self, base_url, dumb=None, opener=None, config=None, **kwargs):
+    def __init__(self, base_url, dumb=None, opener=None, config=None,
+                 username=None, passwd=None, **kwargs):
```
Comment (jelmer) [inline dulwich/client.py]: s/passwd/password/

File: dulwich/client.py [hunk 3]
```diff
@@ -997,13 +997,22 @@ def default_urllib2_opener(config):
...
         if opener is None:
             self.opener = default_urllib2_opener(config)
         else:
             self.opener = opener
+
+        if passwd is None:
+            passwd = ""
+
+        if username:
+            pass_man=urllib2.HTTPPasswordMgrWithDefaultRealm()
```
Comment (jelmer) [inline dulwich/client.py]: Please use spaces around equals signs, per PEP8

File: dulwich/client.py [hunk 4]
```diff
@@ -997,13 +997,22 @@ def default_urllib2_opener(config):
...
+                 username=None, passwd=None, **kwargs):
         self._base_url = base_url.rstrip("/") + "/"
         self.dumb = dumb
         if opener is None:
             self.opener = default_urllib2_opener(config)
         else:
             self.opener = opener
+
+        if passwd is None:
+            passwd = ""
```
Comment (jelmer) [inline dulwich/client.py]: No password specified is not the same as no password; you probably to check for 'username is not None and password is not None' on line 1012.

File: dulwich/client.py [hunk 5]
```diff
@@ -1187,8 +1197,9 @@ def get_transport_and_path_from_url(url, config=None, **kwargs):
         return SSHGitClient(parsed.hostname, port=parsed.port,
                             username=parsed.username, **kwargs), path
     elif parsed.scheme in ('http', 'https'):
-        return HttpGitClient(urlparse.urlunparse(parsed), config=config,
-                **kwargs), parsed.path
+        base_url = urlparse.urlunparse(parsed._replace(netloc=host))
```
Comment (jelmer) [inline dulwich/client.py]: Why not add this logic in HttpGitClient? That would also remove the need to add 'username' and 'password' arguments to it.

**Comment by jelmer** [comment]
Merged with some modifications, thanks :)

**Comment by Israphel** [comment]
is this available on 0.15.0 ?

**Comment by jelmer** [comment]
No, it's not in 0.15.0 yet. It'll be in the next release, which I hope to release in the next couple of days.

**Comment by jelmer** [comment]
0.16.0 is released now with these changes

**Pull request closed without merging**